### ISSUE TITLE
Package request: `verdaccio` npm registry

### ISSUE THREAD

**Issue by roberth** [NixOS/nixpkgs#232762]
**Project description**

Package `verdaccio`, an open source npm registry

I've made a very low effort attempt, but parked this while _somewhat_ blocked by pnpm package support:
 - https://github.com/NixOS/nixpkgs/issues/231513

This could be worked around by hosting a separate npm lock file somewhere, fetching that, and use that in `buildNpmPackage`.

WIP meta snippet:

```nix
  meta = with lib; {
    description = "A simple, zero-config-required local private npm registry";
    # TODO: check with packaging guidelines
    longDescription = ''
      Verdaccio is a simple, zero-config-required local private npm registry. No need for an entire database just to get started! Verdaccio comes out of the box with its own tiny database, and the ability to proxy other registries (eg. npmjs.org), caching the downloaded modules along the way. For those looking to extend their storage capabilities, Verdaccio supports various community-made plugins to hook into services such as Amazon's s3, Google Cloud Storage or create your own plugin.
    '';
    homepage = "https://verdaccio.org";
    license = licenses.mit;
  };
```

**Metadata**

* homepage URL: https://verdaccio.org
* source URL: https://github.com/verdaccio/verdaccio
* license: mit
* platforms: inherit from nodejs?

**Comment by c3n21** [comment]
Any chance to get this merged?

**Comment by c3n21** [comment]
> Afaik there's no PR yet. I just wrote the metadata for the issue in `meta` attr style.
> 
> > I've made a very low effort attempt
> 
> Don't even remember tbh. If it had any value, I would have posted it here or as a draft. I shouldn't have mentioned. I'm sorry.
> 
> EDIT: I'd review it, so I would think chances are pretty good once someone gives it a try. As for me doing the packaging, I'm occupied with other things now.

Oh sorry I didn't notice it was a plain issue

In the next few days perhaps I could try to take a look at it

**Comment by c3n21** [comment]
https://github.com/NixOS/nixpkgs/pull/398441 the package is built but modules caching doesn't seem to work.

**Comment by c3n21** [comment]
@roberth  if you want to review the PR now it's ready

I tested it a bit and it's working for me, but I just began to use Verdaccio so maybe I might miss something

**Comment by c3n21** [comment]
Ok this doesn't seem to be so easy.

In a nutshell, Verdaccio 6.1.2 doesn't use `pnpm` but `yarn`, and more specifically it uses `pnp` for module resolution, which doesn't seem to be supported in nixpkgs (see [my comment](https://github.com/NixOS/nixpkgs/pull/398441#issuecomment-2837603035)) so I can't follow the common pattern in nixpkgs to use yarn-berry to fetch the deps.

The current implementation in the PR uses [this](https://github.com/NixOS/nixpkgs/blob/master/doc/languages-frameworks/javascript.section.md#adding-and-updating-javascript-packages-in-nixpkgs-javascript-adding-or-updating-packages) procedure to package verdaccio, which at least allows us to run it, but it breaks the pipeline because it seems to pull in python2.

Right now I'm waiting for further instructions from people more knowledgeable than me

**Comment by juanpicado** [comment]
>  but yarn, and more specifically it uses pnp for module resolution

If that's a blocker, might have a fix switching to yarn node_linker instead pnp.

**Comment by c3n21** [comment]
> > but yarn, and more specifically it uses pnp for module resolution
> 
> If that's a blocker, might have a fix switching to yarn node_linker instead pnp. Would that help?

Thanks for reaching out!

I already tried by forking the branch 6.x and using `nodeLinker: node-modules` in `.yarnrc.yml` and it does the trick as I don't have to put up any workaround to make it work.

If you want I can open a PR for Verdaccio and force it to use `node-modules`

**Comment by juanpicado** [comment]
I need to do small research first, since might affect Docker image and not sure if is a breaking change. Anyhow future versions 7.x will start using pnpm by default so just maybe work wait a bit.

**Comment by juanpicado** [comment]
@c3n21 I've started removing pnp and pure node_modules from now on, the `verdaccio/verdaccio:6.x-next` should be ready for testing.

**Comment by c3n21** [comment]
> @c3n21 I've started removing pnp and pure node_modules from now on, the `verdaccio/verdaccio:6.x-next` should be ready for testing.

Thanks for telling me :)
